### PR TITLE
Allow using multiple controlled inner blocks refering to the same entity

### LINKED ISSUE #22639: Adding duplicate Template Part removes both from page, can trigger clearing inner blocks in dirty state.
**Describe the bug**
This seems to be a recent regression.  Adding the same template part to a page twice will cause both to be removed from the page.  Navigating through the page content (up and down arrows) may then also trigger the same bug noted in #22638.

**To reproduce**
1. Go to the editor (page editor shown below) with full site editing experiment enabled.
2. Add a template part.
3. Add the same template part again.
4. See that both template parts are removed from the page.  Scroll (up and down arrows) through page content to see some 'funky' behavior and additionally trigger dirty state of the template part to have cleared all of its content (second part likely related to #22638) .

**Expected behavior**
Expected both template parts to remain (previous behavior).  Or if we are iterating to prevent that, at least 1 should remain and a notification should occur.

**Screenshots**
duplicate-template-part

Since this is a recent regression, perhaps @epiqueras, @youknowriad, @noahtallen, @vindl, @ockham, or others may have an idea of where this could originate from?

### PR DESCRIPTION
**Pull request by youknowriad**
closes #22639

This uses a very simple trick to solve the duplicated controller inner blocks bug (duplicate a template part and edit one of them and see it reflected properly).

The main issue was that the block-editor is meant to only work with unique block ids, the block tree can't contain duplications. To solve this, the trick used in this PR is that for each "controlled inner blocks", instead of editing the controlled blocks directly, clone these blocks when there's an "external" change for these blocks (initialization, changing coming from a duplicated block...) and only pass to the editor clones which means the IDs are different. 

Now, this can create infinite loops: an edit in one block, is going to cause changes in the other blocks too. We need to stop  the edits in the second block from triggering back an edit on the original block. Fortunately, this is already handled in useBlockSync by saying: "The first change after an external change is ignored.

Also, there might be a small performance impact: All the other duplicated blocks will remount (all blocks change) on a change happening on one of them. That said, I believe this impact is acceptable for two reasons:

 - The  "cloning" (remounting) only happens for the clones, not the block being currently edited which means it happens only for unselected blocks which means it's async thanks to async mode.
 - The cloning is not performed unless we're in a controlled inner blocks, and this is limitted to template parts and potentially reusable blocks, I don't expect these reusable entities to be very big (like a long post for instance).

### CODE REVIEW

File: packages/block-editor/src/components/provider/use-block-sync.js [hunk 1]
```diff
@@ -113,6 +120,37 @@ export default function useBlockSync( {
 		onChangeRef.current = onChange;
 	}, [ onInput, onChange ] );
 
+	// Determine if blocks need to be reset when they change.
```
Comment (youknowriad) [inline packages/block-editor/src/components/provider/use-block-sync.js:123]: The changes after this line are not necessary (just reordering the hooks). I believe this order makes more sense. "External" then "Internal" changes.

Comment (youknowriad) [reply packages/block-editor/src/components/provider/use-block-sync.js:123]: Actually it seems the reordering has an effect, without it, this PR doesn't work :)

File: packages/block-editor/src/components/provider/use-block-sync.js [hunk 2]
```diff
@@ -96,8 +98,18 @@ export default function useBlockSync( {
 		if ( clientId ) {
 			setHasControlledInnerBlocks( clientId, true );
 			__unstableMarkNextChangeAsNotPersistent();
-			replaceInnerBlocks( clientId, controlledBlocks );
+			// This needs to be called after the setHasControlledInnerBlocks call
```
Comment (mtias) [inline packages/block-editor/src/components/provider/use-block-sync.js:101]: Should it also say why?

Comment (youknowriad) [reply packages/block-editor/src/components/provider/use-block-sync.js:101]: Good catch, this is actually not needed anymore, it's an artifact of one of my earlier explorations.

**Comment by youknowriad** [comment]
Going to merge this. I want to follow-up with a few things here:

 - Try this for the reusable blocks
 - Make an e2e test with two reusable blocks in the same page.

**Comment by gziolo** [comment]
This is very exciting to see, great work 👏

**Comment by youknowriad** [comment]
Santa came was some hours early this year.

**Comment by noahtallen** [comment]
Thanks for getting this fix in @youknowriad! It's very similar to one of my earlier explorations, which would have been way easier had I known about the `cloneBlock` function. 🙈 I agree with you that performance is hopefully not an issue in most cases -- and only an issue with the new behavior allowed by this PR, not anything existing.

**Comment by noisysocks** [comment]
This breaks navigation and widgets because both of those screens maintain a mapping of block client IDs to widget / menu item IDs. Cloning the block changes the client ID which makes the screen unable to tell which widget / menu item that block belongs to.

**Comment by youknowriad** [comment]
> Is there an alternative way to fix this?

Can't think of any to be honest, we tried several approaches with no luck. 

>  block client IDs to widget / menu item IDs

Shouldn't the menuId/widgetId be an attribute of the blocks them selves?

**Comment by noahtallen** [comment]
Possibly we could avoid cloning the block in the case that there are no other things referencing the same entity. (I think now it clones every time?) 

That might solve it for widgets + navigation, unless those screens have a possibility of showing the same entity in more than one place

**Comment by youknowriad** [comment]
@noahtallen true but it seems that we'll be making the screens dependent on an implementation detail. 

It should be made clear whether the "block clientIds" you supply could be changed or not by the editor. I lean towards yes since it's not really the only place where the ids can be swapped. For instance, hitting "Enter" at the end of a paragraph actually changes the id of that paragraph.

**Comment by noisysocks** [comment]
> Shouldn't the menuId/widgetId be an attribute of the blocks them selves?

Wouldn't this mean that _every_ block has to have a `widgetId` attribute?

> @noahtallen true but it seems that we'll be making the screens dependent on an implementation detail.

I disagree that client IDs are an implementation detail. We expose them in several selectors (`getBlocks()`, etc.) and accept them in several selectors and actions (`getBlock()`, `updateBlockAttributes()`, etc.) It's completely reasonable that a developer would use a client ID as a foreign key and so I think that we need to take care to keep them stable.

**Comment by youknowriad** [comment]
> Wouldn't this mean that every block has to have a widgetId attribute?

I'm not really sure how the widgets and navigation screen works tbh to be able to give the right reply here. In my mind (originally), the widget area as a whole is a single "widget" (at least that's how it used to work originally) so it would get a widgetId. If it's different now each block is a widget, then yes, the ones that are widgets should probably get a widget block  wrapper or something referring to the corresponding widget, don't you think? 

> I disagree that client IDs are an implementation detail. We expose them in several selectors (getBlocks(), etc.) and accept them in several selectors and actions (getBlock(), updateBlockAttributes(), etc.) It's completely reasonable that a developer would use a client ID as a foreign key and so I think that we need to take care to keep them stable.

Client IDs are internal  to the block editor store, they are APIs inside the block-editor package but I don't think we ever had the expectation that clientIds don't change from an external API point  of view. In fact, they are considered "temporary" (not persisted) from the start so subject to change.

**Comment by talldan** [comment]
> Shouldn't the menuId/widgetId be an attribute of the blocks them selves?

@youknowriad I had the same thought separately (link to the slack convo - https://wordpress.slack.com/archives/C01D71823PB/p1610003184385900), but there are still a few challenges with that:
- Duplicating blocks would cause duplicate widget ids.
- We could extend all blocks to have a `widgetId` attribute, but that would make them less portable, I haven't tested this, but I'm not sure copy/pasting them to other editors would work, the concern is the extra `widgetId` attribute may cause validation issues.

I don't know if we'd consider adding extra properties to attribute declarations like follows to solve those issues:
```js
"widgetId": {
    "type": "string",
    "isSerializable": "false", // not serialized to HTML.
    "isCopyable": "false" // ignored by `cloneBlock`.
}
```

**Comment by youknowriad** [comment]
- Serializing comment attributes won't be an issue I think, and if you paste it, I believe the attribute might just get removed.
 - Duplicating is an issue, your proposal might work, curious what others would think about it.

**Comment by talldan** [comment]
> Serializing comment attributes won't be an issue I think, and if you paste it, I believe the attribute might just get removed.

You're right, that's good, one less thing to worry about.

**Pull request merged**